File: thor_gestor_de_arquivos_digitais/backend/app/models/admissao.py
from __future__ import annotations

import uuid
from enum import Enum

from sqlalchemy import (
    Boolean,
    BigInteger,
    Date,
    DateTime,
    Enum as SAEnum,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.enums import TipoSuporte


class TipoProcessoAdmissao(str, Enum):
    FECHADO = "FECHADO"
    CONTINUO = "CONTINUO"


class TipoIngressoAdmissao(str, Enum):
    TRANSFERENCIA = "TRANSFERENCIA"
    RECOLHIMENTO = "RECOLHIMENTO"
    DOACAO = "DOACAO"
    AQUISICAO = "AQUISICAO"
    INCORPORACAO = "INCORPORACAO"
    REGULARIZACAO_LEGADO = "REGULARIZACAO_LEGADO"
    OUTRO = "OUTRO"


class StatusProcessoAdmissao(str, Enum):
    ABERTO = "ABERTO"
    EM_NEGOCIACAO = "EM_NEGOCIACAO"
    EM_RECEBIMENTO = "EM_RECEBIMENTO"
    EM_QUARENTENA = "EM_QUARENTENA"
    EM_VALIDACAO = "EM_VALIDACAO"
    PENDENTE_COMPLEMENTACAO = "PENDENTE_COMPLEMENTACAO"
    EM_GERACAO_AIP = "EM_GERACAO_AIP"
    CONCLUIDO = "CONCLUIDO"
    CANCELADO = "CANCELADO"
    REJEITADO = "REJEITADO"


class ResultadoFinalAdmissao(str, Enum):
    ADMITIDO = "ADMITIDO"
    ADMITIDO_COM_RESSALVA = "ADMITIDO_COM_RESSALVA"
    REJEITADO = "REJEITADO"
    CANCELADO = "CANCELADO"


class TipoReuniaoAdmissao(str, Enum):
    NEGOCIACAO_INICIAL = "NEGOCIACAO_INICIAL"
    ALINHAMENTO_TECNICO = "ALINHAMENTO_TECNICO"
    VALIDACAO_SIP = "VALIDACAO_SIP"
    REVISAO_ACORDO = "REVISAO_ACORDO"
    TRATAMENTO_DIVERGENCIA = "TRATAMENTO_DIVERGENCIA"
    HOMOLOGACAO = "HOMOLOGACAO"
    ENCERRAMENTO = "ENCERRAMENTO"
    OUTRO = "OUTRO"


class StatusAcordoAdmissao(str, Enum):
    RASCUNHO = "RASCUNHO"
    EM_ANALISE = "EM_ANALISE"
    ATIVO = "ATIVO"
    SUSPENSO = "SUSPENSO"
    ENCERRADO = "ENCERRADO"


class CanalSubmissao(str, Enum):
    UPLOAD = "UPLOAD"
    API = "API"
    REDE_INTERNA = "REDE_INTERNA"
    MIDIA_REMOVIVEL = "MIDIA_REMOVIVEL"
    ENTREGA_FISICA = "ENTREGA_FISICA"
    IMPORTACAO_SISTEMA = "IMPORTACAO_SISTEMA"
    OUTRO = "OUTRO"


class StatusSessaoSubmissao(str, Enum):
    INICIADA = "INICIADA"
    EM_TRANSFERENCIA = "EM_TRANSFERENCIA"
    RECEBIDA = "RECEBIDA"
    EM_QUARENTENA = "EM_QUARENTENA"
    EM_VALIDACAO = "EM_VALIDACAO"
    VALIDADA = "VALIDADA"
    REJEITADA = "REJEITADA"
    FINALIZADA = "FINALIZADA"
    CANCELADA = "CANCELADA"


class StatusSipAdmissao(str, Enum):
    RECEBIDO = "RECEBIDO"
    EM_QUARENTENA = "EM_QUARENTENA"
    EM_VALIDACAO = "EM_VALIDACAO"
    VALIDADO = "VALIDADO"
    VALIDADO_COM_RESSALVA = "VALIDADO_COM_RESSALVA"
    REJEITADO = "REJEITADO"
    TRANSFORMADO_EM_AIP = "TRANSFORMADO_EM_AIP"


class TipoEventoAdmissao(str, Enum):
    CRIACAO_PROCESSO = "CRIACAO_PROCESSO"
    REUNIAO_ADMISSAO = "REUNIAO_ADMISSAO"
    CRIACAO_VERSAO_ACORDO = "CRIACAO_VERSAO_ACORDO"
    ATIVACAO_ACORDO = "ATIVACAO_ACORDO"
    INICIO_SESSAO = "INICIO_SESSAO"
    RECEBIMENTO_SIP = "RECEBIMENTO_SIP"
    APROVACAO = "APROVACAO"
    REJEICAO = "REJEICAO"
    GERACAO_AIP = "GERACAO_AIP"
    ENCERRAMENTO_PROCESSO = "ENCERRAMENTO_PROCESSO"
    CANCELAMENTO_PROCESSO = "CANCELAMENTO_PROCESSO"


class ResultadoEventoAdmissao(str, Enum):
    SUCESSO = "SUCESSO"
    FALHA = "FALHA"
    ALERTA = "ALERTA"
    PENDENTE = "PENDENTE"
    INFORMATIVO = "INFORMATIVO"


class TipoRelacaoSipAip(str, Enum):
    ORIGEM_TOTAL = "ORIGEM_TOTAL"
    ORIGEM_PARCIAL = "ORIGEM_PARCIAL"
    COMPLEMENTACAO = "COMPLEMENTACAO"
    CORRECAO = "CORRECAO"
    REPROCESSAMENTO = "REPROCESSAMENTO"


class ProcessoAdmissao(Base):
    __tablename__ = "processos_admissao"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    numero_processo: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    titulo: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    descricao: Mapped[str | None] = mapped_column(Text)
    id_instituicao_arquivo: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("instituicao_arquivo.id"), nullable=False, index=True)
    id_entidade_produtora: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("entidades_produtoras.id"), nullable=False, index=True)
    tipo_processo_admissao: Mapped[TipoProcessoAdmissao] = mapped_column(SAEnum(TipoProcessoAdmissao), nullable=False, index=True)
    tipo_ingresso: Mapped[TipoIngressoAdmissao] = mapped_column(SAEnum(TipoIngressoAdmissao), nullable=False, index=True)
    tipo_suporte: Mapped[TipoSuporte] = mapped_column(SAEnum(TipoSuporte), nullable=False, index=True)
    data_inicio: Mapped[Date] = mapped_column(Date, nullable=False, index=True)
    data_fim_prevista: Mapped[Date | None] = mapped_column(Date)
    data_encerramento: Mapped[Date | None] = mapped_column(Date)
    processo_ativo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true", index=True)
    admissoes_recorrentes: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, server_default="false")
    status: Mapped[StatusProcessoAdmissao] = mapped_column(SAEnum(StatusProcessoAdmissao), nullable=False, default=StatusProcessoAdmissao.ABERTO, index=True)
    resultado_final: Mapped[ResultadoFinalAdmissao | None] = mapped_column(SAEnum(ResultadoFinalAdmissao))
    id_descricao_arquivistica: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("registros_descritivos.id"), index=True)
    codigo_classificacao: Mapped[str | None] = mapped_column(String(100))
    codigo_classificacao_descricao: Mapped[str | None] = mapped_column(String(255))
    restricao_acesso: Mapped[str | None] = mapped_column(String(255))
    hipotese_legal_restricao: Mapped[str | None] = mapped_column(String(255))
    volume_estimado: Mapped[str | None] = mapped_column(String(100))
    volume_recebido: Mapped[str | None] = mapped_column(String(100))
    quantidade_unidades_estimadas: Mapped[int | None] = mapped_column(Integer)
    quantidade_unidades_recebidas: Mapped[int | None] = mapped_column(Integer)
    observacoes: Mapped[str | None] = mapped_column(Text)
    parecer_final: Mapped[str | None] = mapped_column(Text)
    criado_por: Mapped[str | None] = mapped_column(String(255))
    atualizado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    atualizado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())

    instituicao_arquivo = relationship("InstituicaoArquivo")
    entidade_produtora = relationship("EntidadeProdutora")
    descricao_arquivistica = relationship("RegistroDescritivo")
    reunioes = relationship("ReuniaoAdmissao", back_populates="processo", cascade="all, delete-orphan")
    acordos = relationship("AcordoAdmissao", back_populates="processo", cascade="all, delete-orphan")
    sessoes = relationship("SessaoSubmissao", back_populates="processo", cascade="all, delete-orphan")
    sips = relationship("SipAdmissao", back_populates="processo", cascade="all, delete-orphan")
    eventos = relationship("EventoAdmissao", back_populates="processo", cascade="all, delete-orphan")


class ReuniaoAdmissao(Base):
    __tablename__ = "reunioes_admissao"
    __table_args__ = (UniqueConstraint("id_processo_admissao", "numero_reuniao", name="uq_reuniao_admissao_numero"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_processo_admissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processos_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    numero_reuniao: Mapped[int] = mapped_column(Integer, nullable=False)
    titulo: Mapped[str] = mapped_column(String(255), nullable=False)
    descricao: Mapped[str | None] = mapped_column(Text)
    tipo_reuniao: Mapped[TipoReuniaoAdmissao] = mapped_column(SAEnum(TipoReuniaoAdmissao), nullable=False, index=True)
    data_reuniao: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False)
    participantes: Mapped[str | None] = mapped_column(Text)
    deliberacoes: Mapped[str | None] = mapped_column(Text)
    pendencias: Mapped[str | None] = mapped_column(Text)
    proximos_passos: Mapped[str | None] = mapped_column(Text)
    ata_documento: Mapped[str | None] = mapped_column(String(500))
    criado_por: Mapped[str | None] = mapped_column(String(255))
    atualizado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    atualizado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())

    processo = relationship("ProcessoAdmissao", back_populates="reunioes")


class AcordoAdmissao(Base):
    __tablename__ = "acordos_admissao"
    __table_args__ = (UniqueConstraint("id_processo_admissao", "numero_versao", name="uq_acordo_admissao_versao"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_processo_admissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processos_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    numero_versao: Mapped[int] = mapped_column(Integer, nullable=False)
    titulo: Mapped[str] = mapped_column(String(255), nullable=False)
    descricao: Mapped[str | None] = mapped_column(Text)
    status: Mapped[StatusAcordoAdmissao] = mapped_column(SAEnum(StatusAcordoAdmissao), nullable=False, default=StatusAcordoAdmissao.RASCUNHO, index=True)
    data_inicio_vigencia: Mapped[Date | None] = mapped_column(Date)
    data_fim_vigencia: Mapped[Date | None] = mapped_column(Date)
    motivo_revisao: Mapped[str | None] = mapped_column(Text)
    regras_empacotamento: Mapped[str | None] = mapped_column(Text)
    regras_nomenclatura: Mapped[str | None] = mapped_column(Text)
    formatos_aceitos: Mapped[str | None] = mapped_column(Text)
    metadados_obrigatorios: Mapped[str | None] = mapped_column(Text)
    requisitos_fixidez: Mapped[str | None] = mapped_column(Text)
    requisitos_representacao: Mapped[str | None] = mapped_column(Text)
    politica_validacao: Mapped[str | None] = mapped_column(Text)
    politica_rejeicao: Mapped[str | None] = mapped_column(Text)
    politica_normalizacao: Mapped[str | None] = mapped_column(Text)
    politica_sigilo: Mapped[str | None] = mapped_column(Text)
    periodicidade_submissao: Mapped[str | None] = mapped_column(String(100))
    observacoes: Mapped[str | None] = mapped_column(Text)
    documento_acordo: Mapped[str | None] = mapped_column(String(500))
    criado_por: Mapped[str | None] = mapped_column(String(255))
    atualizado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    atualizado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())

    processo = relationship("ProcessoAdmissao", back_populates="acordos")
    sessoes = relationship("SessaoSubmissao", back_populates="acordo")


class SessaoSubmissao(Base):
    __tablename__ = "sessoes_submissao"
    __table_args__ = (UniqueConstraint("id_processo_admissao", "numero_sessao", name="uq_sessao_submissao_numero"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_processo_admissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processos_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    id_acordo_utilizado: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("acordos_admissao.id"), nullable=False, index=True)
    numero_sessao: Mapped[int] = mapped_column(Integer, nullable=False)
    titulo: Mapped[str] = mapped_column(String(255), nullable=False)
    descricao: Mapped[str | None] = mapped_column(Text)
    data_inicio: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False)
    data_fim: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True))
    canal_submissao: Mapped[CanalSubmissao] = mapped_column(SAEnum(CanalSubmissao), nullable=False, index=True)
    protocolo_transferencia: Mapped[str | None] = mapped_column(String(255))
    responsavel_envio: Mapped[str | None] = mapped_column(String(255))
    responsavel_recebimento: Mapped[str | None] = mapped_column(String(255))
    tipo_suporte: Mapped[TipoSuporte] = mapped_column(SAEnum(TipoSuporte), nullable=False, index=True)
    volume_informado: Mapped[str | None] = mapped_column(String(100))
    volume_recebido: Mapped[str | None] = mapped_column(String(100))
    quantidade_itens_informada: Mapped[int | None] = mapped_column(Integer)
    quantidade_itens_recebida: Mapped[int | None] = mapped_column(Integer)
    caminho_origem: Mapped[str | None] = mapped_column(String(500))
    caminho_destino_quarentena: Mapped[str | None] = mapped_column(String(500))
    status: Mapped[StatusSessaoSubmissao] = mapped_column(SAEnum(StatusSessaoSubmissao), nullable=False, default=StatusSessaoSubmissao.INICIADA, index=True)
    resultado_validacao: Mapped[str | None] = mapped_column(Text)
    observacoes: Mapped[str | None] = mapped_column(Text)
    criado_por: Mapped[str | None] = mapped_column(String(255))
    atualizado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    atualizado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())

    processo = relationship("ProcessoAdmissao", back_populates="sessoes")
    acordo = relationship("AcordoAdmissao", back_populates="sessoes")
    sips = relationship("SipAdmissao", back_populates="sessao", cascade="all, delete-orphan")


class SipAdmissao(Base):
    __tablename__ = "sips_admissao"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_processo_admissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processos_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    id_sessao_submissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("sessoes_submissao.id", ondelete="CASCADE"), nullable=False, index=True)
    codigo_sip: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    titulo: Mapped[str] = mapped_column(String(255), nullable=False)
    descricao: Mapped[str | None] = mapped_column(Text)
    tipo_sip: Mapped[TipoSuporte] = mapped_column(SAEnum(TipoSuporte), nullable=False, index=True)
    status: Mapped[StatusSipAdmissao] = mapped_column(SAEnum(StatusSipAdmissao), nullable=False, default=StatusSipAdmissao.RECEBIDO, index=True)
    data_recebimento: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False)
    estrutura_original: Mapped[str | None] = mapped_column(Text)
    caminho_armazenamento_temporario: Mapped[str | None] = mapped_column(String(500))
    manifesto_arquivos: Mapped[str | None] = mapped_column(Text)
    algoritmo_hash: Mapped[str | None] = mapped_column(String(50))
    hash_global: Mapped[str | None] = mapped_column(String(255))
    tamanho_bytes: Mapped[int | None] = mapped_column(BigInteger)
    quantidade_arquivos: Mapped[int | None] = mapped_column(Integer)
    quantidade_unidades_fisicas: Mapped[int | None] = mapped_column(Integer)
    resultado_validacao: Mapped[str | None] = mapped_column(Text)
    observacoes: Mapped[str | None] = mapped_column(Text)
    criado_por: Mapped[str | None] = mapped_column(String(255))
    atualizado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    atualizado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())

    processo = relationship("ProcessoAdmissao", back_populates="sips")
    sessao = relationship("SessaoSubmissao", back_populates="sips")


class RelacaoSipAip(Base):
    __tablename__ = "relacoes_sip_aip"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_sip: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("sips_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    id_unidade_acondicionamento: Mapped[int] = mapped_column(ForeignKey("unidades_acondicionamento.id"), nullable=False, index=True)
    tipo_relacao: Mapped[TipoRelacaoSipAip] = mapped_column(SAEnum(TipoRelacaoSipAip), nullable=False, default=TipoRelacaoSipAip.ORIGEM_TOTAL)
    observacoes: Mapped[str | None] = mapped_column(Text)
    criado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())

    sip = relationship("SipAdmissao")
    unidade_acondicionamento = relationship("UnidadeAcondicionamento")


class EventoAdmissao(Base):
    __tablename__ = "eventos_admissao"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    id_processo_admissao: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processos_admissao.id", ondelete="CASCADE"), nullable=False, index=True)
    id_sessao_submissao: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("sessoes_submissao.id", ondelete="SET NULL"), index=True)
    id_sip: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("sips_admissao.id", ondelete="SET NULL"), index=True)
    id_unidade_acondicionamento: Mapped[int | None] = mapped_column(ForeignKey("unidades_acondicionamento.id", ondelete="SET NULL"), index=True)
    tipo_evento: Mapped[TipoEventoAdmissao] = mapped_column(SAEnum(TipoEventoAdmissao), nullable=False, index=True)
    descricao: Mapped[str] = mapped_column(Text, nullable=False)
    resultado: Mapped[ResultadoEventoAdmissao] = mapped_column(SAEnum(ResultadoEventoAdmissao), nullable=False, default=ResultadoEventoAdmissao.INFORMATIVO, index=True)
    agente: Mapped[str | None] = mapped_column(String(255))
    data_evento: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), index=True)
    detalhe_tecnico: Mapped[str | None] = mapped_column(Text)
    evidencia: Mapped[str | None] = mapped_column(String(500))
    criado_por: Mapped[str | None] = mapped_column(String(255))
    criado_em: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())

    processo = relationship("ProcessoAdmissao", back_populates="eventos")

File: thor_gestor_de_arquivos_digitais/frontend/lib/api/admissao.ts
import { apiRequest } from "@/lib/api/client";
import type { TipoSuporte } from "@/types/domain";

export type TipoProcessoAdmissao = "FECHADO" | "CONTINUO";
export type TipoIngressoAdmissao = "TRANSFERENCIA" | "RECOLHIMENTO" | "DOACAO" | "AQUISICAO" | "INCORPORACAO" | "REGULARIZACAO_LEGADO" | "OUTRO";
export type StatusProcessoAdmissao = "ABERTO" | "EM_NEGOCIACAO" | "EM_RECEBIMENTO" | "EM_QUARENTENA" | "EM_VALIDACAO" | "PENDENTE_COMPLEMENTACAO" | "EM_GERACAO_AIP" | "CONCLUIDO" | "CANCELADO" | "REJEITADO";
export type ResultadoFinalAdmissao = "ADMITIDO" | "ADMITIDO_COM_RESSALVA" | "REJEITADO" | "CANCELADO";
export type TipoReuniaoAdmissao = "NEGOCIACAO_INICIAL" | "ALINHAMENTO_TECNICO" | "VALIDACAO_SIP" | "REVISAO_ACORDO" | "TRATAMENTO_DIVERGENCIA" | "HOMOLOGACAO" | "ENCERRAMENTO" | "OUTRO";
export type StatusAcordoAdmissao = "RASCUNHO" | "EM_ANALISE" | "ATIVO" | "SUSPENSO" | "ENCERRADO";
export type CanalSubmissao = "UPLOAD" | "API" | "REDE_INTERNA" | "MIDIA_REMOVIVEL" | "ENTREGA_FISICA" | "IMPORTACAO_SISTEMA" | "OUTRO";
export type StatusSessaoSubmissao = "INICIADA" | "EM_TRANSFERENCIA" | "RECEBIDA" | "EM_QUARENTENA" | "EM_VALIDACAO" | "VALIDADA" | "REJEITADA" | "FINALIZADA" | "CANCELADA";
export type StatusSipAdmissao = "RECEBIDO" | "EM_QUARENTENA" | "EM_VALIDACAO" | "VALIDADO" | "VALIDADO_COM_RESSALVA" | "REJEITADO" | "TRANSFORMADO_EM_AIP";
export type TipoEventoAdmissao = "CRIACAO_PROCESSO" | "REUNIAO_ADMISSAO" | "CRIACAO_VERSAO_ACORDO" | "ATIVACAO_ACORDO" | "INICIO_SESSAO" | "RECEBIMENTO_SIP" | "APROVACAO" | "REJEICAO" | "GERACAO_AIP" | "ENCERRAMENTO_PROCESSO" | "CANCELAMENTO_PROCESSO";
export type ResultadoEventoAdmissao = "SUCESSO" | "FALHA" | "ALERTA" | "PENDENTE" | "INFORMATIVO";

export type ProcessoAdmissao = {
  id: string;
  numero_processo: string;
  titulo: string;
  descricao?: string | null;
  id_instituicao_arquivo: string;
  id_entidade_produtora: string;
  tipo_processo_admissao: TipoProcessoAdmissao;
  tipo_ingresso: TipoIngressoAdmissao;
  tipo_suporte: TipoSuporte;
  data_inicio: string;
  data_fim_prevista?: string | null;
  data_encerramento?: string | null;
  processo_ativo: boolean;
  admissoes_recorrentes: boolean;
  status: StatusProcessoAdmissao;
  resultado_final?: ResultadoFinalAdmissao | null;
  id_descricao_arquivistica?: string | null;
  codigo_classificacao?: string | null;
  codigo_classificacao_descricao?: string | null;
  restricao_acesso?: string | null;
  hipotese_legal_restricao?: string | null;
  volume_estimado?: string | null;
  volume_recebido?: string | null;
  quantidade_unidades_estimadas?: number | null;
  quantidade_unidades_recebidas?: number | null;
  observacoes?: string | null;
  parecer_final?: string | null;
  nome_instituicao_arquivo?: string | null;
  nome_entidade_produtora?: string | null;
  titulo_descricao_arquivistica?: string | null;
  criado_por?: string | null;
  atualizado_por?: string | null;
  criado_em: string;
  atualizado_em: string;
};

export type ProcessoAdmissaoPayload = Omit<ProcessoAdmissao, "id" | "criado_em" | "atualizado_em" | "nome_instituicao_arquivo" | "nome_entidade_produtora" | "titulo_descricao_arquivistica">;
export type ProcessoAdmissaoFilters = Partial<Pick<ProcessoAdmissao, "numero_processo" | "titulo" | "id_entidade_produtora" | "tipo_processo_admissao" | "tipo_ingresso" | "tipo_suporte" | "status"> & { q: string; processo_ativo: string; data_inicio_de: string; data_inicio_ate: string }>;
export type ProcessoAdmissaoPage = { items: ProcessoAdmissao[]; total: number; limit: number; offset: number };

export type ReuniaoAdmissao = { id: string; id_processo_admissao: string; numero_reuniao: number; titulo: string; descricao?: string | null; tipo_reuniao: TipoReuniaoAdmissao; data_reuniao: string; participantes?: string | null; deliberacoes?: string | null; pendencias?: string | null; proximos_passos?: string | null; ata_documento?: string | null; criado_em: string; atualizado_em: string };
export type AcordoAdmissao = { id: string; id_processo_admissao: string; numero_versao: number; titulo: string; descricao?: string | null; status: StatusAcordoAdmissao; data_inicio_vigencia?: string | null; data_fim_vigencia?: string | null; motivo_revisao?: string | null; regras_empacotamento?: string | null; regras_nomenclatura?: string | null; formatos_aceitos?: string | null; metadados_obrigatorios?: string | null; requisitos_fixidez?: string | null; politica_validacao?: string | null; politica_rejeicao?: string | null; observacoes?: string | null; documento_acordo?: string | null; criado_em: string; atualizado_em: string };
export type SessaoSubmissao = { id: string; id_processo_admissao: string; id_acordo_utilizado: string; numero_sessao: number; titulo: string; descricao?: string | null; data_inicio: string; data_fim?: string | null; canal_submissao: CanalSubmissao; tipo_suporte: TipoSuporte; status: StatusSessaoSubmissao; responsavel_envio?: string | null; responsavel_recebimento?: string | null; volume_informado?: string | null; volume_recebido?: string | null; observacoes?: string | null; criado_em: string; atualizado_em: string };
export type SipAdmissao = { id: string; id_processo_admissao: string; id_sessao_submissao: string; codigo_sip: string; titulo: string; descricao?: string | null; tipo_sip: TipoSuporte; status: StatusSipAdmissao; data_recebimento: string; caminho_armazenamento_temporario?: string | null; algoritmo_hash?: string | null; hash_global?: string | null; tamanho_bytes?: number | null; quantidade_arquivos?: number | null; quantidade_unidades_fisicas?: number | null; resultado_validacao?: string | null; observacoes?: string | null; criado_em: string; atualizado_em: string };
export type EventoAdmissao = { id: string; id_processo_admissao: string; id_sessao_submissao?: string | null; id_sip?: string | null; id_unidade_acondicionamento?: number | null; tipo_evento: TipoEventoAdmissao; descricao: string; resultado: ResultadoEventoAdmissao; agente?: string | null; data_evento: string; detalhe_tecnico?: string | null; evidencia?: string | null; criado_em: string };

function queryString(params: Record<string, string | number | undefined | null>) {
  const searchParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && String(value).trim() !== "") searchParams.set(key, String(value));
  });
  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export function listProcessosAdmissao({ limit = 20, offset = 0, filters = {} }: { limit?: number; offset?: number; filters?: ProcessoAdmissaoFilters } = {}) {
  return apiRequest<ProcessoAdmissaoPage>(`/admissao/processos${queryString({ limit, offset, ...filters })}`);
}
export function getProcessoAdmissao(id: string) { return apiRequest<ProcessoAdmissao>(`/admissao/processos/${id}`); }
export function createProcessoAdmissao(payload: ProcessoAdmissaoPayload) { return apiRequest<ProcessoAdmissao>("/admissao/processos", { method: "POST", body: JSON.stringify(payload) }); }
export function updateProcessoAdmissao(id: string, payload: Partial<ProcessoAdmissaoPayload>) { return apiRequest<ProcessoAdmissao>(`/admissao/processos/${id}`, { method: "PUT", body: JSON.stringify(payload) }); }
export function deleteProcessoAdmissao(id: string) { return apiRequest<void>(`/admissao/processos/${id}`, { method: "DELETE" }); }

export function listReunioesAdmissao(processoId: string) { return apiRequest<ReuniaoAdmissao[]>(`/admissao/processos/${processoId}/reunioes`); }
export function createReuniaoAdmissao(processoId: string, payload: Partial<ReuniaoAdmissao>) { return apiRequest<ReuniaoAdmissao>(`/admissao/processos/${processoId}/reunioes`, { method: "POST", body: JSON.stringify(payload) }); }
export function listAcordosAdmissao(processoId: string) { return apiRequest<AcordoAdmissao[]>(`/admissao/processos/${processoId}/acordos`); }
export function createAcordoAdmissao(processoId: string, payload: Partial<AcordoAdmissao>) { return apiRequest<AcordoAdmissao>(`/admissao/processos/${processoId}/acordos`, { method: "POST", body: JSON.stringify(payload) }); }
export function ativarAcordoAdmissao(id: string) { return apiRequest<AcordoAdmissao>(`/admissao/acordos/${id}/ativar`, { method: "POST" }); }
export function novaVersaoAcordoAdmissao(id: string) { return apiRequest<AcordoAdmissao>(`/admissao/acordos/${id}/nova-versao`, { method: "POST" }); }
export function listSessoesSubmissao(processoId: string) { return apiRequest<SessaoSubmissao[]>(`/admissao/processos/${processoId}/sessoes`); }
export function createSessaoSubmissao(processoId: string, payload: Partial<SessaoSubmissao>) { return apiRequest<SessaoSubmissao>(`/admissao/processos/${processoId}/sessoes`, { method: "POST", body: JSON.stringify(payload) }); }
export function finalizarSessaoSubmissao(id: string) { return apiRequest<SessaoSubmissao>(`/admissao/sessoes/${id}/finalizar`, { method: "POST" }); }
export function listSipsProcesso(processoId: string) { return apiRequest<SipAdmissao[]>(`/admissao/processos/${processoId}/sips`); }
export function createSipAdmissao(sessaoId: string, payload: Partial<SipAdmissao>) { return apiRequest<SipAdmissao>(`/admissao/sessoes/${sessaoId}/sips`, { method: "POST", body: JSON.stringify(payload) }); }
export function validarSipAdmissao(id: string) { return apiRequest<SipAdmissao>(`/admissao/sips/${id}/validar`, { method: "POST" }); }
export function rejeitarSipAdmissao(id: string) { return apiRequest<SipAdmissao>(`/admissao/sips/${id}/rejeitar`, { method: "POST" }); }
export function listEventosAdmissao(processoId: string) { return apiRequest<EventoAdmissao[]>(`/admissao/processos/${processoId}/eventos`); }
export function createEventoAdmissao(processoId: string, payload: Partial<EventoAdmissao>) { return apiRequest<EventoAdmissao>(`/admissao/processos/${processoId}/eventos`, { method: "POST", body: JSON.stringify(payload) }); }
